### DISCUSSION TITLE
ComfyUI needs some default maths and number nodes - prove me wrong.

### DISCUSSION THREAD

**Discussion by Ferniclestix** [comfyanonymous/ComfyUI#1229] [Ideas]
![ints](https://github.com/comfyanonymous/ComfyUI/assets/128430314/9138b7d6-8b4a-43a3-852a-416d61cf7357)

This is what happens when basic needs are not met by a program.
Multiple custom nodes are providing their own solution to an issue and its messing up my interface with too many options.

The need not being met is that of a re-routable multi use set of math nodes. this is because of comfyUI's strange attachment to the contextual 'primitive' nodes. which are not flexible enough for the intended purpose.  (no cross connecting certain inputs between node types for example in the case of width and height for image and latent upscalers despite both accepting int values, no rerouting due to contextual requirement not being passed by reroute nodes ect.)

Imo, we need an actual set of multi use math nodes that can be used as a standard connector set for custom nodes, its a basic missing framework required for custom nodes to make use of comfyUI's systems.

and it needs to be in comfyUI by default so we dont have soo many copies of those same nodes.

**Comment by ltdrdata** [comment]
I agree with this totally.

**Comment by toyxyz** [comment]
I totally agree too!

**Comment by Ruffy314** [comment]
I think the issue is that people don't know how to use the *Primitive* node. The primitive will turn into an INT / FLOAT / STRING node, if connected to an INT / FLOAT / STRING input. Until then it looks useless, unfortunately, and the displayed message looks like some default text for a node template. 

A math node could be useful though.

**Reply by Ferniclestix** [reply]
Connect the width of a latent upscale to a primitive node. Then connect the width of an image upscale to the same node.

Reroute a primitive node.

We know how to use it, they don't work properly.
Due to the contextual nature of inputs to primitive nodes, you can't connect certain things to the same primitive node that should be using the same variables. and combining inputs is a major thing prim nodes are useful for.

I would actually use prim nodes if they worked, the double click an input to get a prim node is amazing.

**Comment by Feynt** [comment]
Came to confirm, I'm not able to perform math on any primitives.  Trying to go from ksampler to ksampler (advanced of course), I'm providing a primitive for the stop/start steps respectively.  It's obviously working as it's accepted on both sampler nodes.  But I can't _also_ take the output of that same primitive and route it through a math node to increment it by 1 so it starts at the right step on the next node.  It's like it's not actually an INT, or the INT used by ksampler (advanced) nodes with widgets converted to inputs which expect numbers is not consistent with the input on the math nodes (any of them, official or community added).

But, and here's the kicker:  If I make a new primitive node, connect it first to an INT accepting node (I added a custom unary INT operator node), and _then_ also connect it to the steps node on one of the ksampler (advanced) nodes I have, it works.

**Comment by Pos13** [comment]
100%.

I like to share my workflows, and I constantly feel the lack of default math node. Forcing the recepient of your workflow to install WAS Node Suite just for single "divide by 2" operation is not nice.

"Constant Int", "Constant Float" and "Constant String" are also desirable. Primitives are great, but they are very tricky to work with any nodes containing flexible "*" inputs. Solvable for default reroutes, but not for custom nodes (highways, senders, scripts, etc.)

**Comment by Pos13** [comment]
Update: it's _not_ solvable for default reroutes: #2866
